Sources/URLRouting/Fragment.swift

```diff
@@ -0,0 +1,74 @@
+import Parsing
+
+/// Parses a request's fragment subcomponent.
+public struct Fragment<FragmentParser: Parser>: Parser where FragmentParser.Input == Substring {
+    
+  @usableFromInline
+  let fragmentParser: FragmentParser
+  
+  @inlinable
+  public init(@ParserBuilder build: () -> FragmentParser) {
+    self.fragmentParser = build()
+  }
+  
+  /// Initializes a fragment parser from a substring conversion.
+  ///
+  /// Useful for parsing a request fragment in its entirety, for example as a String.
+  ///
+  /// ```swift
+  /// Fragment(.string)
+  /// ```
+  ///
+  /// - Parameter fragmentConversion: A conversion that transforms fragment into some other type.
+  @inlinable
+  public init<C>(_ fragmentConversion: C)
+  where FragmentParser == Parsers.MapConversion<Parsers.ReplaceError<Rest<FragmentParser.Input>>, C> {
+    self.fragmentParser = Rest().replaceError(with: .init()).map(fragmentConversion)
+  }
+  
+  /// Initializes a fragment parser from a substring conversion and a parser of the conversion's output.
+  ///
+  /// Useful for transforming a request fragment in its entirety, and then parsing on the transformed value.
+  ///
+  /// For example, if your server responds with a query string in the URL's fragment, you could create a
+  /// "fields" conversion that transforms a `Substring` into a `URLRequestData.Fields` type. With
+  /// this in place you could parse the individual ``Field`` elements in the following query-like URL
+  /// fragment:
+  ///
+  /// `/search#query=Blob`
+  ///
+  /// ```swift
+  /// Fragment(.fields) {
+  ///   Field("query", default: "")
+  /// }
+  /// ```
+  ///
+  /// - Parameters:
+  ///   - fragmentConversion: A conversion that transforms the fragment into some other type.
+  ///   - parsers: A parser builder that can parse the conversion's output.
+  @inlinable
+  public init<C: Conversion, P: Parser>(
+    _ fragmentConversion: C,
+    @ParserBuilder with parsers: () -> P
+  )
+  where FragmentParser == Parsers.Pipe<Parsers.MapConversion<Parsers.ReplaceError<Rest<Substring>>, C>, P> {
```

Comment:
I think it makes sense to have an alternate interface that takes a parser builder (just like `Body` does), but I think we should limit the number of "mapping" operators like this to higher-level constructs like `Route`, `Parse`, and `ParsePrint`, especially since there are more overloads to consider.

Effectively, your example could be something like:

```swift
ParsePrint(
  .fields {
    Field("query", default: "")
  }
) {
  Fragment()
}

// or:

Fragment().map(.fields {
  Field("query", default: "")
})
```